resources/ChatServer.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 3/31/08               \" DATE 
.Dt ChatServer 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm ChatServer,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner 



src/com/presence/chat/ChatServer.java
//
//  ChatServer.java
//  ChatServer
//
//  Created by John McKisson on 3/31/08.
//  Copyright (c) 2008 __MyCompanyName__. All rights reserved.
//
package com.presence.chat;

import java.net.*;
import java.io.*;
import java.util.*;
import java.util.logging.*;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.awt.event.*;

import org.jboss.netty.channel.Channel;

import com.presence.chat.commands.*;
import com.presence.chat.plugin.*;
import com.presence.chat.socket.*;

import static com.presence.chat.ANSIColor.*;

public class ChatServer {
	
	static ChatServer instance;
	
	public static ChatServer getServer() {
		return instance;
	}

	static List<ChatClient> clients = new ArrayList<ChatClient>();
	static Hashtable<String, ChatRoom> rooms = new Hashtable<String, ChatRoom>();
	
	static Hashtable<String, CommandEntry> commands;
	
	javax.swing.Timer spamTimer = null;
	javax.swing.Timer shutdownTimer = null;

	
	static {
		commands = new Hashtable<String, CommandEntry>();
	
		//Level 0 commands
		commands.put("ca",		new CommandEntry(new CMDChatAll(), 0));
		commands.put("compact", new CommandEntry(new CMDCompact(), 0));
		commands.put("help",	new CommandEntry(new CMDHelp(), 0));
		commands.put("join",	new CommandEntry(new CMDJoinRoom(), 0));
		//commands.put("plog",	new CMDPersonalLog());
		commands.put("log",		new CommandEntry(new CMDLog(), 0));
		commands.put("motd",	new CommandEntry(new CMDMOTD(), 0));
		commands.put("plugins",	new CommandEntry(new CMDPlugins(), 0));
		commands.put("pmsg",	new CommandEntry(new CMDChatPrivate(), 0));
		commands.put("who",		new CommandEntry(new CMDWho(), 0));
		commands.put("stats",	new CommandEntry(new CMDStats(), 0));
	
		//Level 1 commands
		commands.put("ccreate",	new CommandEntry(new CMDCreateRoom(), 1));
		commands.put("chpw",	new CommandEntry(new CMDPassword(), 1));
		//info Get detailed info on a user
		//setip Enable/reset ipauth feature
		
		//Level 2 commands
		commands.put("ignore",		new CommandEntry(new CMDIgnore(), 2));
		//commands.put("gca",		new CMDGlobalChatAll());
		
		//Level 3 commands
		commands.put("auth",	new CommandEntry(new CMDAuth(), 3));
		commands.put("fjoin",	new CommandEntry(new CMDForceJoin(), 3));
		commands.put("silent",	new CommandEntry(new CMDSilent(), 3));
		//remove Force removal of a channel
		
		//Level 4 commands
		commands.put("add",		new CommandEntry(new CMDAdd(), 4));
		commands.put("kick",	new CommandEntry(new CMDKick(), 4));
		
		commands.put("setlvl",	new CommandEntry(new CMDSetLevel(), 4));
		commands.put("setpw",	new CommandEntry(new CMDSetPassword(), 4));
		//ban Ban a username or IP address
		//unban
		//banlist
		//gag Gag a user for specified time
		
		//Level 5 commands
		commands.put("accounts",new CommandEntry(new CMDAccounts(), 5));
		commands.put("del",		new CommandEntry(new CMDDelete(), 5));
		commands.put("shutdown",new CommandEntry(new CMDShutdown(), 5));
		//commands.put("wl",		new CommandEntry(new CMDWriteLog(), 5));
	}
	
	public static void addCommand(String str, Command cmd, int level) {
		commands.put(str, new CommandEntry(cmd ,level));
	}
	

	public static final String VERSION = "jChatServ by Humera, " + ChatServer.class.getPackage().getImplementationVersion();
	
	public static String USAGE_STRING = ChatPrefs.getName() + " chats to you, 'Usage: /chat " + ChatPrefs.getName() + " %s'";
	
	//Use the same ByteBuffer for all channels.  A single thread is servicing all the channels
	//so there is no danger of concurrent access
	//public static ByteBuffer BUF = ByteBuffer.allocateDirect(40960);	//40k buffer

	
	//Selector mainSelector;
	
	boolean keepRunning = true;
	
	public PluginManager pManager;
	
	ServerStats stats;

	public ChatServer() throws Exception {
	
		instance = this;
		
		stats = new ServerStats();
		
		//Create default chat room
		ChatRoom mainRoom = new ChatRoom("main");
		mainRoom.setDestroyable(false);
		
		rooms.put("main", mainRoom);

		AccountManager.loadAccounts();
		
		pManager = new PluginManager();
		pManager.installPlugins();
		
		initSpamTimer();
		
		TelnetServer telnetServer = new TelnetServer(ChatPrefs.getPort());
		
		while (keepRunning) {
			try {
				Thread.sleep(500);
			} catch (InterruptedException e) {
			}
		}
		
		Logger.getLogger("global").info("Serving shutting down");
		
		telnetServer.shutdown();
		
		AccountManager.saveAccounts();
	}
	
	
	void initSpamTimer() {
		//Set up spam timer to reset spam counters every 2 seconds
		spamTimer = new javax.swing.Timer(2000, new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				for (ChatClient client : clients)
					client.spamThresh = 0;
					
				spamTimer.start();
			}
		});
		spamTimer.setRepeats(false);
		spamTimer.start();
	}
	
	/*
	void mainLoop() throws Exception {
		//Allocate an unbound server socket channel
		ServerSocketChannel serverChannel = ServerSocketChannel.open();
		
		ServerSocket serverSocket = serverChannel.socket();
		
		mainSelector = Selector.open();
		
		try {
			serverSocket.bind(new InetSocketAddress(ChatPrefs.getPort()));
		} catch (BindException e) {
			Logger.getLogger("global").severe("Address already bound, restarting in 5 seconds");
			serverSocket = null;
			Thread.sleep(5000);
			return;
		}
		
		serverChannel.configureBlocking(false);
		
		//Register selector with ssc
		serverChannel.register(mainSelector, SelectionKey.OP_ACCEPT);
		
		Logger.getLogger("global").info("Server ready, listening on port " + ChatPrefs.getPort());
		
		while (keepRunning) {

			//I'm using select without a timeout here to avoid additional overhead
			//The selector is an instance member so we're able to wakeup() externally
			int n = mainSelector.select();
			
			if (n == 0) continue;
			
			Iterator it = mainSelector.selectedKeys().iterator();
			
			while (it.hasNext()) {
				SelectionKey key = (SelectionKey)it.next();
				
				//Invalid key, connection dropped or something
				if (!key.isValid()) {
					SocketChannel socketChannel = (SocketChannel)key.channel();
		
					ChatClient theClient = clients.get(socketChannel);
					
					if (theClient == null) {
						Logger.getLogger("global").warning("Something bad happened trying to read data from a client that doesnt exist!");
						socketChannel.close();
					} else {
						Logger.getLogger("global").warning("Closing socket with invalid key");
						theClient.disconnect();
					}
				}
				
				//New connection
				if (key.isAcceptable()) {
					ServerSocketChannel server = (ServerSocketChannel)key.channel();
					
					SocketChannel channel = server.accept();
					
					registerClient(mainSelector, channel, SelectionKey.OP_READ);
				}
				
				//New data
				if (key.isReadable()) {
					readDataFromSocket(key);
				}
				
				//Done with this key
				it.remove();
			}
		}
		
		serverChannel.close();
	}
	*/
	
	/*
	void registerClient(Selector selector, SocketChannel channel, int ops) throws Exception {
		if (channel == null) return;
		
		channel.configureBlocking(false);
		
		channel.register(selector, ops);	//Regsiter with selector
		
		clients.put(channel, new ChatClient(channel));
	}
	*/
	
	/**
	 * Closes a clients socket, removes them from the client list, and logs their disconnect
	 */
	static void disconnectClient(ChatClient client) {
		Channel sock = client.getSocket();
		
		if (sock != null) {
			
			sock.close();
			

			clients.remove(sock);
		}
		
		Logger.getLogger("global").info(String.format("%s disconnected", client.getName()));
	}

	/*
	void readDataFromSocket(SelectionKey key) throws Exception {
		SocketChannel socketChannel = (SocketChannel)key.channel();
		
		ChatClient theClient = clients.get(socketChannel);
		
		if (theClient == null) {
			Logger.getLogger("global").warning("Something bad happened trying to read data from a client that doesnt exist!");
			socketChannel.close();
			return;
		}
		
		BUF.clear();
		
		try {
			int numBytesRead = socketChannel.read(BUF);
			
			if (numBytesRead == -1) {
				//Close channel on EOF, this invalidates the key
				
				Logger.getLogger("global").warning("EOF on socket");
				theClient.disconnect();
			
			} else {
				BUF.flip();
				
				theClient.processIncomingData();
			}

		} catch (IOException e) {
			//Ok, we don't know what happened here...  This seems to be disconnecting clients a lot
			Logger.getLogger("global").warning("Client Exception");
			e.printStackTrace();
			
			//theClient.disconnect();
		}
	}
	*/
	
	
	public static ServerStats getStats() {
		return instance.stats;
	}
	
	
	/*
	 * Return a string representation of the bytes in the global buffer
	 */
	/*
	public static String bufToString() {
		byte[] bbuf = new byte[BUF.remaining()];
		BUF.get(bbuf);
		return new String(bbuf);
	}
	
	public static byte[] buffer() {
		byte[] b = new byte[BUF.remaining()];
		BUF.get(b);
		return b;
	}
	*/
	
	
	/**
	 * Process a server command from a client
	 * @param myClient Client that sent the personal msg
	 * @param content Contents of the msg
	 */
	static final Matcher matcherChatPrivate = Pattern.compile("^(.*) chats to you, '(.*)'$").matcher("");
	 
	public static void processCommand(ChatClient client, String content) {
		
		matcherChatPrivate.reset(content.trim());
		
		if (!matcherChatPrivate.find()) {
			Logger.getLogger("global").info("Invalid private chat syntax from " + client.getName());
			System.out.println(content + NRM);
			
			return;
		}
		
		String clientName = matcherChatPrivate.group(1).toLowerCase();

		//Make sure incoming chat name matches the name of the client sending the command
		if (!client.getName().toLowerCase().equals(clientName)) {
			Logger.getLogger("global").info(String.format("\tClient names do not match!  %s / %s", client.getName(), matcherChatPrivate.group(1)));
			return;
		}
			
		//Treat this input as a password if the client is not yet authenticated
		if (!client.isAuthenticated()) {
			//User supplied a password, check if it matches
				
			ChatAccount account = client.getAccount();
			
			//System.out.printf("comparing '%s' to '%s'\n", matcher.group(2), account.getPassword());
			
			String cryptedPassword = JCrypt.crypt("", matcherChatPrivate.group(2));
			
			if (cryptedPassword.equals(account.getPassword())) {
				client.setAuthenticated(true, "Password Auth");
				
			} else {
				client.sendChat(String.format("%s chats to you, 'Incorrect Password'", ChatPrefs.getName()));
				
				client.getProtocol().closeSocket();	
			}
			
			return;
		}
		
		//log.info(String.format("CMD from %s: '%s'", matcher.group(1), matcher.group(2)));
		
		//Split off the actual command and arguments to it
		String[] args = matcherChatPrivate.group(2).split(" ", 2);
		
		CommandEntry cmdEntry = getCommand(args[0]);
		
		if (cmdEntry == null || cmdEntry.level > client.getAccount().getLevel()) {
			client.sendChat(String.format("%s chats to you, 'You idiot!  There is no <%s> command!'", ChatPrefs.getName(), args[0]));
			return;
		} else {
			instance.stats.cmds++;
			
			cmdEntry.cmd.execute(client, args);
		}
	}
	
	
	/**
	 * Get a named command entry
	 * @param cmd Command name
	 * @return CommandEntry named by name, null if no entry exists
	 */
	public static CommandEntry getCommand(String cmd) {
		return commands.get(cmd);
	}
	
	public static Hashtable<String, CommandEntry> getCommands() {
		return commands;
	}
	
	
	/**
	 * @return true if the specified person is online, false otherwise
	 */
	public static boolean checkOnlineName(String testName, String skip) {
		Iterator<ChatClient> it = clients.iterator();
		
		while (it.hasNext()) {
			ChatClient c = it.next();
			
			String name = c.getName();
			
			if (name != null && name.equalsIgnoreCase(testName) && !name.equalsIgnoreCase(skip)) {
				return true;
			}
		}
		
		return false;
	}
	
	
	public static ChatClient getClientByName(String name) {
		
		for (Iterator<ChatClient> it = getClients().iterator() ; it.hasNext() ;) {
			ChatClient targ = it.next();
			
			if (targ.getName() == null) continue;
			
			if (targ.getName().equalsIgnoreCase(name))
				return targ;
				
		}
		
		return null;
	}
	
	
	public static List<ChatClient> getClients() {
		return clients;
	}
	
	public static ChatRoom getRoom(String name) {
		return rooms.get(name.toLowerCase());
	}
	
	public static Hashtable<String, ChatRoom> getRooms() {
		return rooms;
	}
	
	
	/**
	 * Echo a message to everyone in all rooms
	 * @param msg Message to be displayed
	 */
	public static void echo(String msg) {
		Enumeration<ChatRoom> e = rooms.elements();
		
		while (e.hasMoreElements()) {
			ChatRoom room = e.nextElement();
			
			room.echo(msg, null);
		}
	}
	
	
	public static void shutdown(int seconds) {
		instance._shutdown(seconds);
	}
	
	void _shutdown(int seconds) {

		if (shutdownTimer != null)
			shutdownTimer = null;
			
		shutdownTimer = new javax.swing.Timer(seconds * 1000, new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				echo(String.format("%s[%s%s%s] Server %sRestart%s",
					RED, WHT, ChatPrefs.getName(), RED, YEL, RED));
			
				keepRunning = false;
		
				//mainSelector.wakeup();
			}
		});
		
		shutdownTimer.setRepeats(false);
		shutdownTimer.start();
	}
	
	
	static void parseArgs(String args[]) {
		int n = 0;
		
		while (args.length >= n + 1) {
		
			if (args[n + 1] != null /*&& !args[n + 1].startsWith("-")*/) {
			
				if (args[n].equals("-p")) {
					ChatPrefs.setPort(Integer.parseInt(args[n + 1]));
					
				} else if (args[n].equals("-n")) {
					ChatPrefs.setName(args[n + 1]);
				}

			} else
				break;
			
			n += 2;
		}
	}
	

	public static void printStringBytes(String str) {
		StringBuilder strBuf = new StringBuilder();
		byte[] b = str.getBytes();
		for (int i = 0; i < b.length; i++)
			strBuf.append(String.format("%02X ", b[i]));
			
		System.out.println(strBuf.toString() + "\n");
	}
	


    public static void main (String args[]) throws Exception {
		
		//Setup logging facilities
		LogManager logManager = LogManager.getLogManager();
        logManager.reset();
		
		Logger log = Logger.getLogger("global");
		
		log.setUseParentHandlers(false);
		
		ChatLogFormatter clf = new ChatLogFormatter();
		
		// log file max size 10K, 3 rolling files, append-on-open
        Handler fileHandler = new FileHandler("log/syslog%g.log", 10000, 6, true);
        fileHandler.setFormatter(clf);
		log.addHandler(fileHandler);
		
		Handler consoleHandler = new ConsoleHandler();
		consoleHandler.setFormatter(clf);
		log.addHandler(consoleHandler);
		
		//Redirect stdout and stderr to our logger
        System.setOut(new PrintStream(new LoggingOutputStream(log, StdOutErrLevel.STDOUT), true));
        System.setErr(new PrintStream(new LoggingOutputStream(log, StdOutErrLevel.STDERR), true));
		
		//Parse arguments
		parseArgs(args);
		
		// now show stderr stack trace going to logger
		/*
        try {
            throw new RuntimeException("Test");
        } catch (Exception e) {
            e.printStackTrace();
        }
		*/

		
		try {
			new ChatServer();
		} catch (Exception e) {
			e.printStackTrace();
		}
    }
}
